src/Microsoft.AspNetCore.Routing.Abstractions/RouteValueDictionary.cs

```diff
@@ -439,42 +439,107 @@ internal abstract class Storage
 
         internal class ListStorage : Storage
         {
-            internal readonly List<KeyValuePair<string, object>> _inner;
+            private KeyValuePair<string, object>[] _items;
+            private int _size;
 
             public ListStorage()
             {
-                _inner = new List<KeyValuePair<string, object>>();
+                _items = new KeyValuePair<string, object>[0];
             }
 
             public ListStorage(int capacity)
             {
-                _inner = new List<KeyValuePair<string, object>>(capacity);
+                _items = new KeyValuePair<string, object>[capacity];
             }
 
             public ListStorage(ListStorage other)
             {
-                // Perf: Don't call the copy constructor, that would box the enumerator.
-                _inner = new List<KeyValuePair<string, object>>(other._inner.Capacity);
-                for (var i = 0; i < other._inner.Count; i++)
+                _items = new KeyValuePair<string, object>[other.Capacity];
+                for (var i = 0; i < other.Count; i++)
                 {
-                    _inner.Add(other._inner[i]);
+                    this.Add(other[i]);
                 }
             }
 
-            public override int Count => _inner.Count;
+            public int Capacity
```

Comment:
Why does this property need to exist, can't you just do all of this in `EnsureCapacity`?